docs/design/proposals/direct-blk-device-assignment.md

```diff
@@ -0,0 +1,246 @@
+* **NodeGetVolumeStats** -- Since the container runtime mounts the filesystem, it needs to forward the request to the container runtime.
+* **NodeExpandVolume** -- The CSI cannot implement this API alone. It needs to propagate the new size to the container runtime so that the filesystem metadata can be refreshed accordingly.
+* **NodeStageVolume/NodeUnStageVolume** -- Invoke a kata-runtime handler to clean up the stale mount information.
+
+### Kata-runtime changes
+The host agent (CSI node driver, as an example) will call into kata-runtime. Kata-runtime behaves as the API for 
+sandbox-level management interactions. In order for the CSI node driver to communicate with the kata-runtime, 
+we propose augmenting kata-runtime to include a filesystem, `fs`, command as follows:
+
+```
+kata-runtime fs —help
+NAME:
+kata-runtime fs [subcommand] [options]— interact with mounted filesystem volume
+
+SUBCOMMANDS:
+ - add-blk-mount-info <mountinfo> -- add the mount info of a volume to the Kata runtime.
+ - reset-blk-mount-info <host-blk-device-name> –- clean up the mount info directory under kata runtime
+ - get-stats <host-blk-device-name> — show status and usage details for the mounted filesystem backed by the given block device, <host-blk-device-name>.
+ - resize <host-blk-device-name> <size> — update the mounted volume to <size> for the mounted filesystem backed by the given block device, <host-blk-device-name>.
+
+```
+The mountinfo object is defined as follows:
+```
+type MountInfo struct {
+    // Device: source device path for this volume
+    Device string json:"device,omitempty"
+    // FsType: filesystem that needs to be used to mount the storage inside the VM
+    FsType string json:"fsType,omitempty"
+    // metadata field for easy extension
+    Metadata map[string]string json:"metadata,omitempty"
+    // Options: additional options that might be needed to mount the storage filesystem
+    Options []string json:"options,omitempty"
+}
+```
+
+#### Quick notes on implementation
+
+Instead of the CSI node driver writing the mount info into a `csiPlugin.json` file under the volume root, 
+as described in the original proposal. Here we propose that the CSI node driver passes the mount information to 
+the kata-runtime through a new kata-runtime command called `add-blk-mount-info`. The kata-runtime then write the mount 
+information to a `“`mount-info.json`”` file in a hardcoded location (e.g. `/run/kata-containers/shared/direct-block-devices/[block_dev_name]/`).
+
+When a Kata Pod sandbox is started, the containerd-kata-shim-v2 checks whether there is any mount identified 
+by its source path in the container specification that has a mount info file in the `direct-block-devices` directory. 
+If it exists, the containerd-kata-shim-v2 parses the mount info file and propagates it to the kata agent to carry out the mount. 
+Lastly,the containerd-kata-shim-v2 creates a file named by the sandbox id under the `direct-block-devices/[block_dev_name]/` 
```

Comment:
Question: is the file(`a file named by the sandbox id`) empty and does only the file name has its meaning?

Reply:
Exactly.